onnxruntime/core/providers/cuda/nn/rms_norm.cc

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/nn/rms_norm.h"
+#include "core/providers/cuda/nn/layer_norm.h"
+#include "core/providers/cuda/nn/layer_norm_impl.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cpu/nn/layer_norm_helper.h"
+#include <vector>
+
+namespace onnxruntime {
+namespace cuda {
+
+// RMSNorm uses LayerNorm kernel, which only supports X and scale both
+// being the same data type.
+#define REGISTER_KERNEL_TYPED(T, V)                                                                 \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(RMSNormalization, kOnnxDomain, 23, T##_##V, kCudaExecutionProvider, \
+                                (*KernelDefBuilder::Create())                                       \
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())          \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<V>()),         \
+                                RMSNorm<T, float, V>);
```

Comment:
Could replace float by `std::conditional<std::is_same_v<T, double>::value, double, float>::type`, then we only need define one macro.

Reply:
Added a typealis MeanVarType which conditionally choose between double and float depending on both T and V.